src/CAN/can_mapping.c

```diff
@@ -44,10 +49,52 @@ float canmapping_extract_value(uint64_t raw_data, const CANMapping *mapping)
                         raw_value = swap_uint32(raw_value);
                         break;
                     default:
+                        /* Should never be mapping anything larger than 4 bytes */
+                        panic(PANIC_CAUSE_UNREACHABLE);
                         break;
                 }
         }
-        return (float)raw_value;
+
+        /* convert type */
+        switch (mapping->type) {
+            case CANMappingType_unsigned:
+                    return (float)raw_value;
+        	case CANMappingType_signed:
+                    if (length <= 8) {
+                        return (float)*((int8_t*)&raw_value);
+                    }
+                    else if (length <= 16){
+                        return (float)*((int16_t*)&raw_value);
+                    }
+                    else {
+                        return (float)*((int32_t*)&raw_value);
+                    }
+            case CANMappingType_IEEE754:
+                    return *((float*)&raw_value);
+            case CANMappingType_sign_magnitude:
+                    /**
+                     *  sign-magnitude is used in cases where there's a sign bit
+                     *  and an absolute value indicating magnitude.
+                     *  e.g. BMW E46 steering angle sensor
+                     **/
+                    {
+                            uint32_t sign;
+                            if (length <= 8) {
+                                    sign = 1 << 7;
```

Comment:
Ok, simplified algorithm where we just calculate the sign based on the length of the mapping.  it works.  Thanks!

case CANMappingType_sign_magnitude:
                    /**
                     *  sign-magnitude is used in cases where there's a sign bit
                     *  and an absolute value indicating magnitude.
                     *  e.g. BMW E46 steering angle sensor
                     **/
                    {
                            uint32_t sign = 1 << (length - 1);
                            return raw_value < sign ? (float)raw_value : -(float)(raw_value & (sign - 1));
                    }